pkg/controller/atlasproject/maintenancewindow.go

```diff
@@ -0,0 +1,157 @@
+			return result
+		}
+		return workflow.OK()
+	}
+
+	if windowSpecified(windowSpec) {
+		ctx.Log.Debugw("Checking if window needs update")
+		windowInAtlas, result := getInAtlas(ctx.Client, projectID)
+		if !result.IsOk() {
+			return result
+		}
+
+		if windowInAtlas.DayOfWeek == 0 || windowInAtlas.HourOfDay == nil ||
+			*windowInAtlas.HourOfDay != windowSpec.HourOfDay || windowInAtlas.DayOfWeek != windowSpec.DayOfWeek {
+			ctx.Log.Debugw("Creating or updating window")
+			// We set startASAP to false because the operator takes care of calling the API a second time if both
+			// startASAP and the new maintenance timeslots are defined
+			if result := createOrUpdateInAtlas(ctx.Client, projectID, windowSpec.WithStartASAP(false)); !result.IsOk() {
+				return result
+			}
+		} else if *windowInAtlas.AutoDeferOnceEnabled != windowSpec.AutoDefer {
+			// If autoDefer flag is different in Atlas, and we haven't updated the window previously, we toggle the flag
+			ctx.Log.Debugw("Toggling autoDefer")
+			if result := toggleAutoDeferInAtlas(ctx.Client, projectID); !result.IsOk() {
+				return result
+			}
+		}
+	}
+
+	if windowSpec.StartASAP {
+		ctx.Log.Debugw("Starting maintenance ASAP")
+		// To avoid any unexpected behavior, we send a request to the API containing only the StartASAP flag,
+		// although the API should ignore other fields in that case
+		if result := createOrUpdateInAtlas(ctx.Client, projectID, project.NewMaintenanceWindow().WithStartASAP(true)); !result.IsOk() {
+			return result
+		}
+		// Nothing else should be done after sending a StartASAP request
+		return workflow.OK()
+	}
+
+	if windowSpec.Defer {
+		ctx.Log.Debugw("Deferring scheduled maintenance")
+		if result := deferInAtlas(ctx.Client, projectID); !result.IsOk() {
+			return result
+		}
+		// Nothing else should be done after deferring
+		return workflow.OK()
+	}
+
+	return workflow.OK()
+}
+
+func isEmpty(i int) bool {
+	return i == 0
+}
+
+func isEmptyWindow(window project.MaintenanceWindow) bool {
+	return isEmpty(window.DayOfWeek) && isEmpty(window.HourOfDay) && !window.StartASAP && !window.Defer && !window.AutoDefer
+}
+
+func windowSpecified(window project.MaintenanceWindow) bool {
+	return !isEmpty(window.DayOfWeek)
+}
+
+func maxOneFlag(window project.MaintenanceWindow) bool {
+	return !(window.StartASAP && window.Defer)
+}
+
+// validateMaintenanceWindow performs validation of the Maintenance Window. Note, that we intentionally don't validate
+// that hour of day and day of week are in the bounds - this will be done by Atlas.
+func validateMaintenanceWindow(window project.MaintenanceWindow) error {
+	if isEmptyWindow(window) || (windowSpecified(window) && maxOneFlag(window)) {
+		return nil
+	}
+	return errors.New(`
+		projectMaintenanceWindow must respect the following constraints, or be empty :
+			1) dayOfWeek must be specified, hourOfDay is 0 by default, autoDeferral is false by default
+			2) only one of (startASAP, defer) is true
+	`)
```

Comment:
This multiline comment looks great in the code but will be ugly in the state. Better to change this message so it looks fine in the state.